File: cpp/src/mapping.cpp
#include <SDL.h>
#include <array>
#include <cmath>
#include <stdexcept>
#include <iostream>
#include <vector>
#include "mapping.h"


Mapping::Mapping() = default;

Mapping::Mapping(int width, int height, float scale) {
    this->width = width;
    this->height = height;
    this->scale = scale;

    if (SDL_Init(SDL_INIT_VIDEO) < 0) {
        SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "SDL could not initialize! SDL_Error: %s\n", SDL_GetError());
        throw std::runtime_error("SDL could not initialize!");
    }

    // Crear una ventana
    SDL_Window* window = SDL_CreateWindow("SDL2 Prime Spiral", SDL_WINDOWPOS_UNDEFINED,
                                          SDL_WINDOWPOS_UNDEFINED, this->width, this->height,
                                          SDL_WINDOW_SHOWN);
    this->window = window;
    if (window == nullptr) {
        SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Window could not be created! SDL_Error: %s\n", SDL_GetError());
        SDL_Quit();
        throw std::runtime_error("Window could not be created!");
    }

    // Crear un renderizador
    SDL_Renderer* renderer = SDL_CreateRenderer(this->window, -1, SDL_RENDERER_ACCELERATED);
    this->renderer = renderer;
    if (renderer == nullptr) {
        SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Renderer could not be created! SDL_Error: %s\n", SDL_GetError());
        SDL_DestroyWindow(window);
        SDL_Quit();
        throw std::runtime_error("Renderer could not be created!");
    }
}

std::array<int, 2> Mapping::mapToScreen(std::array<int, 2> point) {
    int x = point[0];
    int y = point[1];
    int newX = x * this->scale + this->width / 2;
    int newY = -y * this->scale + this->height / 2;
    return {(int) newX, (int) newY};
}

std::array<int, 2> Mapping::screenToMap(std::array<int, 2> point) {
    int x = point[0];
    int y = point[1];
    int newX = (x - this->width / 2) / this->scale;
    int newY = (-y + this->height / 2) / this->scale;
    return {(int) newX, (int) newY};
}

std::array<int, 2> Mapping::rotateVector(std::array<int, 2> point, float angle) {
    int x = point[0];
    int y = point[1];
    float newX = x * cos(angle) - y * sin(angle);
    float newY = x * sin(angle) + y * cos(angle);
    return {(int) newX, (int) newY};
}

void Mapping::drawPoint(std::array<int, 2> point) {
    std::array<int, 2> screenPoint = this->mapToScreen(point);
    SDL_RenderDrawPoint(this->renderer, screenPoint[0], screenPoint[1]);
}

void Mapping::drawMultiplePoints(std::vector<std::array<int, 2>> points, int number_of_points) {
    SDL_RenderDrawPoints(this->renderer, (SDL_Point*) &points[0], number_of_points);
}

void Mapping::setColors(std::array<int, 3> prime_color, std::array<int, 3> background_color) {
    this->prime_color = prime_color;
    this->background_color = background_color;
    SDL_SetRenderDrawColor(this->renderer, this->background_color[0], this->background_color[1], this->background_color[2], 255);    
    SDL_RenderClear(this->renderer);
    SDL_SetRenderDrawColor(this->renderer, this->prime_color[0], this->prime_color[1], this->prime_color[2], 255);
}

void Mapping::update() {
    SDL_RenderPresent(this->renderer);
}

void Mapping::quit() {
    SDL_DestroyRenderer(this->renderer);
    SDL_DestroyWindow(this->window);
    SDL_Quit();
}

int Mapping::getWidth() {
    return this->width;
}

int Mapping::getHeight() {
    return this->height;
}

File: python/mapping.py
import pygame as pg
import numpy as np


class Mapping:
    def __init__(self, screen_size, map_scale):
        self.screen_size = screen_size
        self.screen = pg.display.set_mode(self.screen_size)
        self.map_scale = map_scale
        self.map_origin = np.array(screen_size) / 2

        self.screen.fill((51, 51, 51))

        pg.init()


    def map_to_screen(self, map_point):
        # Map a point from the map coordinate system to the screen coordinate system
        # map_point: [x, y] in map coordinates
        # screen_point: [x, y] in screen coordinates
        newX = map_point[0] * self.map_scale + self.map_origin[0]
        newY = -map_point[1] * self.map_scale + self.map_origin[1]
        return np.array([newX, newY]).astype(int)
    

    def screen_to_map(self, screen_point):
        # Map a point from the screen coordinate system to the map coordinate system
        # screen_point: [x, y] in screen coordinates
        # map_point: [x, y] in map coordinates
        newX = (screen_point[0] - self.map_origin[0]) / self.map_scale
        newY = (-screen_point[1] + self.map_origin[1]) / self.map_scale
        return np.array([newX, newY])
    

    def rotate_vector(self, vector, angle):
        # Rotate a vector by a given angle
        # vector: [x, y]
        # angle: angle in radians
        # rotated_vector: [x, y]
        rotation_matrix = np.array([[np.cos(angle), -np.sin(angle)], [np.sin(angle), np.cos(angle)]])
        rotated_vector = np.dot(rotation_matrix, vector)
        return rotated_vector
    

    def draw_point(self, point, color, radius):
        # Draw a point on the screen
        # point: [x, y] in map coordinates
        # color: (r, g, b)
        # radius: radius in pixels
        screen_point = self.map_to_screen(point)
        pg.draw.circle(self.screen, color, screen_point, radius)